### PR TITLE
Add support for IPv6 gaming

### PR DESCRIPTION
**Pull request by Hagb**

### CODE REVIEW

File: src/Common/Packet.cpp [hunk 1]
```diff
@@ -2,6 +2,7 @@
 // Created by PinkySmile on 01/10/2022.
 //
 
+#include <cstdint>
```
Comment (PinkySmile) [inline src/Common/Packet.cpp:5]: I'm assuming you had compiler error in these files?

File: src/SokuMod/Connection.cpp [hunk 2]
```diff
@@ -13,6 +13,7 @@ std::mutex logMutex;
 #include <Exceptions.hpp>
 #include "Connection.hpp"
 #include "getPublicIp.hpp"
+#include "ipv6map_extern.hpp"
```
Comment (PinkySmile) [inline src/SokuMod/Connection.cpp:16]: Where is that file? It's not included in the PR

Comment (Hagb) [reply src/SokuMod/Connection.cpp:16]: ummm i forgot to add them.

File: src/Common/Packet.cpp (hunk 1 above)
Comment (Hagb) [reply src/Common/Packet.cpp:5]: no error here iirc. probably editor added them automatically when i used uint16_t in this file

File: src/SokuMod/Connection.cpp [hunk 3]
```diff
@@ -309,8 +318,23 @@ bool Connection::_handlePacket(const Lobbies::PacketGameStart &packet, size_t &s
...
+	if (this->onConnectRequest){
+		const char * ip = packet.ip;
+		unsigned short port = packet.port;
+		char ipv6[sizeof(packet.ipv6)+1];
+		char ipv6MappedIpv4[16];
+		if (isIpv6Available() && packet.port6){
+			strncpy(ipv6, packet.ipv6, sizeof(packet.ipv6));
+			ipv6[sizeof(ipv6)-1] = '\0';
+			printf("p1 ipv6 is: %s\n", ipv6);
+			if (mapIpv6toIpv4(ipv6, ipv6MappedIpv4, sizeof(ipv6MappedIpv4))){
```
Comment (PinkySmile) [inline src/SokuMod/Connection.cpp:330]: How does that work? Can you convert any IPv6 to IPv4 somehow!? I assume it's just the IPv6 mod doing some magic behind the scene?

**Review by PinkySmile — COMMENTED** [review]
I like the idea. How much does it help to play without port forward/autopunch? Also, I think you should talk to CC about it because I assume it completely breaks the hostlist. I think the stats and replay gathering of SokuReplays is really essential to help new players get better so I'd rather have it work with it if possible.

File: src/SokuMod/Connection.cpp (hunk 3 above)
Comment (Hagb) [reply src/SokuMod/Connection.cpp:330]: yeah. it is ipv6 mod that does that.
the ipv6 mod hookes network functions and converts virtual ipv4 addresses to ipv6 ones

**Comment by Hagb** [comment]
> How much does it help to play without port forward/autopunch?

Depends on how common players have IPv6 access.

> Also, I think you should talk to CC about it because I assume it completely breaks the hostlist.

yes. If p1 and p2 play via IPv6, hostlist will not work. A workaround at this time is posting the ip (v4) of p2 instead of p1 to hostlist, but it doesn't work if p2 doesn't have autopunch or port forward working.

File: src/SokuMod/Connection.cpp (hunk 3 above)
Comment (Hagb) [reply src/SokuMod/Connection.cpp:330]: > Can you convert any IPv6 to IPv4 somehow!?

Not "real" IPv4 addresses. IPv6 mod maintains maps between them and real IPv6 addresses.

**Comment by Hagb** [comment]
https://github.com/Gegel85/SokuLobbies/blob/6ec4ae32294b07c22de112a89c7978cf7ee5b048/src/Common/Packet.hpp#L87-L98

Is the `data`  member of `PacketOlleh` reserved for future use? If so, I think we can put `PROTOCOL_VERSION` of server (for possibility of backward compatibility) and IPv4 address of client there.

**Comment by PinkySmile** [comment]
The PROTOCOL_VERSION is already checked by the server by the time it receives the OPCODE_HELLO (if they mismatch it will reply with a OPCODE_KICKED instead of OPCODE_OLLEH). This is reserved for true packet incompatibility like opcode changes, change in the size of packets, .... The data field is reserved for lobby customization when it will be a thing. The server will send the layout of the lobby to the client so it can render everything properly.

**Comment by Hagb** [comment]
> The PROTOCOL_VERSION is already checked by the server by the time it receives the OPCODE_HELLO (if they mismatch it will reply with a OPCODE_KICKED instead of OPCODE_OLLEH).

Knowing the protocol version of a (older) server, the client would be possible to be compatible with the server if we need, by turning off some features or something else. Though it might be over-engineering at this time.

> The data field is reserved for lobby customization when it will be a thing.

To pass the addresses of clients to them, should we include the address in `data` or use a new type of packet for it?

**Comment by Hagb** [comment]
> Knowing the protocol version of a (older) server, the client would be possible to be compatible with the server if we need, by turning off some features or something else. Though it might be over-engineering at this time.

In more detail, this would be possible if the client also knew the protocol version of the server and it were the end with newer protocol that rejected the mismatched connection.
In the cases of the client with newer protocol version, if a client is compatible with the server, the client could connect to the server with compatible packets; if the client is newer than and incompatible with the server (i.e. we did nothing extra for the backward compatibility), the client could just simply end the connection.
The cases of newer server are similar.
We don't need to keep the backward compatibility if we don't need it, but we could just make it possible to do that in the future when we do need it.
Sorry for off-topic in this PR.

**Comment by Hagb** [comment]
> > Also, I think you should talk to CC about it because I assume it completely breaks the hostlist.
>
> yes. If p1 and p2 play via IPv6, hostlist will not work. A workaround at this time is posting the ip (v4) of p2 instead of p1 to hostlist, but it doesn't work if p2 doesn't have autopunch or port forward working.

There are also similar (the same in fact) problems on spectating IPv6 games. A spectator without the access to IPv6 is not be able to spectate an IPv6 game via the address of p1 -- even if the spectator can connect to p1 via IPv4, p1 will send [an IPv6 version](https://github.com/Hagb/IPv6Map/blob/master/soku-protocol-extension.md#v6_soku_redirect) of [`REDIRECT` packet](https://github.com/delthas/touhou-protocol-docs/blob/master/protocol_123.md#packet-0x08-redirect) to the spectator and then spectating will fail because the spectator doesn't know the IPv4 address of p2. One possible workaround is:

1. P2 joining a host with `HostPort` in `SokuLobbies.ini` as port, so that if p2 has port forward configured, spectators could connect to p2 via `HostIP` of p2. 
2. P2 reports its addresses and port(s) to the lobby server after a game starts, so that server can tell spectator the addresses of p2 instead of the ones of p1. Spectating might still fail, depending on whether spectator can access p2.
3. (before hostlist server is updated to support spectating via IPv6) P2 puts itself to the hostlist server after an _IPv6_ game starts, so that though the hostlist server fails to spectate via p1 it is _possible_ to spectate p2, depending on whether it can access p2 (via IPv4).

Seems that the workaround makes no new problem (not worse than before). @Gegel85 Could you tell me what you think about it?

**Comment by PinkySmile** [comment]
The main issue is that finding the P2 bound port is annoying. It may be different from the port bound when hosting and I don't think that would really work. I know there is a relay system in the base game so using that would help since the game actively acts as a relay. I think ultimately the best solution is to just do nothing and wait for Konni to be updated so that its relay can be use instead with no drawback.
I made a mod that allows to spectate offline games so it would theoretically also be able to handle that on the side but that's a very dirty solution.

**Comment by Hagb** [comment]
Sorry for my bad English. My expression at my previous comments were probably confusing.

> The main issue is that finding the P2 bound port is annoying. It may be different from the port bound when hosting and I don't think that would really work.

It is possible to custom the p2 port. Like `clientport` configured `configex123.ini`. (I think there may be extra work needed to be done to find the memory address to set it?)
If you meant the port in NAT, then Autopunch should be able to map for us if installed and supported.

> I know there is a relay system in the base game so using that would help since the game actively acts as a relay.

Yeah. According to [the protocol document](https://github.com/delthas/touhou-protocol-docs), this relay system only redirects spectating to p2 or other peers, and punches for spectating using the similar mechanism like Autopunch. Some users may behind a symmetric NAT, who could not be punched by other soku peers or Autopunch afaik.

When they are p2, spectating via the ip of p1 will fail, because p1 redirects the spectating to p2 and then the spectator will not be able to connect to p2 even with punching. It could happen in IPv4. I think Konni doesn't either work in this case.

For those users, port forward is the only way they host or be spectated in IPv4. It is one of the reasons why I suggested to give the ip of p2 instead of p1 to spectators includijg Konni, so that port forward could work when spectator tries to connect to p2. (It is more than only a temporary workaround for IPv6 cases.)

**Comment by Hagb** [comment]
> For those users, port forward is the only way they host or be spectated in IPv4. It is one of the reasons why I suggested to give the ip of p2 instead of p1 to spectators includijg Konni, so that port forward could work when spectator tries to connect to p2. (It is more than only a temporary workaround for IPv6 cases.)

When I said "the ip of p2" here, I meant the real ip of p2 if p2's `HostIP` in `SokuLobbies.ini` is empty, or `HostIP` value if it is not empty.

**Comment by PinkySmile** [comment]
I merged it manually instead

**Pull request closed without merging**